### PR TITLE
Draft: STM32 periph clk: centralize conditional code

### PR DESCRIPTION
**Pull request by Enoch247**
### Contribution description

This is a work in progress. I wanted to put it out there to get some feedback. See review comments for areas where feedback is especially requested.

### Testing procedure

TBD

### Issues/PRs references

This work supports my work to port RIOT to the STM32H7 family.

### CODE REVIEW

File: cpu/stm32/cpu_common.c [hunk 1]
```diff
@@ -75,6 +75,27 @@ uint32_t periph_timer_clk(bus_t bus)
     return periph_apb_clk(bus) * apbmul[bus];
 }
 
+void periph_clk_en2(const periph_t *periph)
+{
+    assert(periph);
+
+    const int irq_state = irq_disable();
```
Comment (Enoch247) [inline cpu/stm32/cpu_common.c:82]: the existing `periph_clk_en()` implementation (I believe) is not thread safe. So I have added critical sections to this implementation. Is my thinking correct that this is needed?

File: cpu/stm32/include/periph/cpu_common.h [hunk 2]
```diff
@@ -113,6 +113,12 @@ typedef enum {
 #endif
 } bus_t;
 
+typedef struct periph_t
```
Comment (Enoch247) [inline cpu/stm32/include/periph/cpu_common.h:116]: oops :)

```suggestion
typedef struct
```

File: cpu/stm32/include/periph/cpu_common.h [hunk 3]
```diff
@@ -113,6 +113,12 @@ typedef enum {
...
+{
+    volatile uint32_t *en_reg;
+    uint32_t en_mask;
+} periph_t;
```
Comment (Enoch247) [inline cpu/stm32/include/periph/cpu_common.h:120]: thoughts on the name of this `sturct`?

File: cpu/stm32/include/periph/cpu_timer.h [hunk 4]
```diff
@@ -44,11 +44,25 @@ extern "C" {
...
+    .en_mask = RCC_APB1ENR_TIM2EN,
+    #elif defined(RCC_APB1ENR1_TIM2EN)
+    .en_reg = &RCC->APB1ENR1,
+    .en_mask = RCC_APB1ENR1_TIM2EN,
+    #elif defined(RCC_MC_APB1ENSETR_TIM2EN)
+    .en_reg = &RCC->MC_APB1ENSETR,
+    .en_mask = RCC_MC_APB1ENSETR_TIM2EN,
+    #endif
+};
+
```
Comment (Enoch247) [inline cpu/stm32/include/periph/cpu_timer.h:59]: Is this the right file to define this in?

File: cpu/stm32/include/periph/cpu_timer.h [hunk 5]
```diff
@@ -44,11 +44,25 @@ extern "C" {
...
 /**
  * @brief   Timer configuration
  */
 typedef struct {
     TIM_TypeDef *dev;       /**< timer device */
+    const periph_t *rcc_dev;
```
Comment (Enoch247) [inline cpu/stm32/include/periph/cpu_timer.h:65]: Is there a better name for this pointer?

File: cpu/stm32/include/periph/cpu_timer.h [hunk 6]
```diff
@@ -44,11 +44,25 @@ extern "C" {
...
     uint32_t max;           /**< maximum value to count to (16/32 bit) */
     uint32_t rcc_mask;      /**< corresponding bit in the RCC register */
     uint8_t bus;            /**< APBx bus the timer is clock from */
```
Comment (Enoch247) [inline cpu/stm32/include/periph/cpu_timer.h:68]: These will go away when this work is complete.

File: cpu/stm32/include/periph/cpu_timer.h (hunk 5 above)
Comment (kfessel) [reply cpu/stm32/include/periph/cpu_timer.h:65]:
```suggestion
    const rcc_register_t *rcc_register;
```

File: cpu/stm32/include/periph/cpu_common.h (hunk 3 above)
Comment (kfessel) [reply cpu/stm32/include/periph/cpu_common.h:120]: the name seems a bit generic, at least the usage in this PR seems like `rcc_register_t` or more specific

File: cpu/stm32/include/periph/cpu_common.h [hunk 7]
```diff
@@ -113,6 +113,17 @@ typedef enum {
...
+/**
+ * @brief STM32 peripherial base type.
+ *
+ * This structure contains things that are common to all STM32 peripherals.
+ */
+typedef struct
+{
+    volatile uint32_t *en_reg;      /**< RCC enable reg */
+    uint32_t en_mask;               /**< RCC enable reg mask */
+} periph_base_t;
```
Comment (kfessel) [inline cpu/stm32/include/periph/cpu_common.h:125]: While every/most periph need to be clocked via the rcc this is certainly not the base of that peripheral, but more like a "RCC configuration" of said periph. (`rcc_config_t` or `periph_clock_config_t`)

If the base of an periph would be pointed at somewhere, I would expect is to point to that periphs base adress eg: for timer 1 to TIM1_CR1 (add an offset to that base and you find the other registers)

**Comment by Enoch247** [comment]
I now believe that this approach was flawed. I have created a PR with an alternate approach (#20609), which is better in my opinion.

**Pull request closed without merging**